src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpConnector.java

```diff
@@ -184,12 +184,17 @@ URLConnection connect(URL originalUrl, Function<URL, ImmutableMap<String, String
                     || code == 505) {  // Server refuses to support version quoth RFC7231 § 6.6.6
           // This is a permanent error so we're not going to retry.
           readAllBytesAndClose(connection.getErrorStream());
+          if (code == 404 || code == 410) {
+            // For Not Found, we throw a separate unrecoverable exception so that callers can
+            // distinguish between the resource being not found and the server being unavailable.
+            throw new FileNotFoundException(describeHttpResponse(connection));
+          }
           throw new UnrecoverableHttpException(describeHttpResponse(connection));
         } else {
           // However we will retry on some 5xx errors, particularly 500 and 503.
           throw new IOException(describeHttpResponse(connection));
         }
-      } catch (UnrecoverableHttpException e) {
+      } catch (UnrecoverableHttpException | FileNotFoundException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)